grails-app/services/org/pih/warehouse/putaway/PutawayService.groovy

```diff
@@ -38,37 +40,25 @@ class PutawayService {
     GrailsApplication grailsApplication
 
     def getPutawayCandidates(Location location) {
-        List putawayItems = []
-        List<Location> internalLocations = locationService.getInternalLocations(location,
-                [ActivityCode.RECEIVE_STOCK] as ActivityCode[])
-
         List binLocationEntries = productAvailabilityService.getAvailableQuantityOnHandByBinLocation(location)
 
-        internalLocations.each { internalLocation ->
-            List putawayItemsTemp = binLocationEntries.findAll {
-                it.binLocation == internalLocation
-            }
-            if (putawayItemsTemp) {
-                putawayItemsTemp = putawayItemsTemp.collect {
-
-                    // FIXME Removed because it was causing very slow performance - determine if this is even necessary
-                    List<AvailableItem> availableItems = []
-
-                    PutawayItem putawayItem = new PutawayItem()
-                    putawayItem.putawayStatus = PutawayStatus.READY
-                    putawayItem.product = it.product
-                    putawayItem.inventoryItem = it.inventoryItem
-                    putawayItem.currentFacility = location
-                    putawayItem.currentLocation = it.binLocation
-                    putawayItem.putawayFacility = null
-                    putawayItem.putawayLocation = null
-                    putawayItem.availableItems = availableItems
-                    putawayItem.quantity = it.quantity
-                    return putawayItem
-                }
-                putawayItems.addAll(putawayItemsTemp)
+        List<PutawayItem> putawayItems = binLocationEntries.inject ([], { putawayItems,  binLocationEntry ->
+            if (binLocationEntry.binLocation.supports(ActivityCode.RECEIVE_STOCK)) {
```

Comment:
There are two tables, location_type_supported_activities and location_supported_activities. 

Aside: I just wrote a larger description of how supported activities work, but then realized that does not pertain to your question so I'm going to use self-restraint and refrain from posting :stuck_out_tongue: But if you'd like to see more context, I'll post it somewhere more useful.

Reply:
Note it would be better for the initial query (persumably against product_availability) to limit the inventory items returned based on the supported activities associated with the "bin location", but I suspect the query for that would be more complicated due to the inheritance relationship between the two supported actitivities tables. That's why I think we're pulling out everything and then filtering. But if we could move the "supports" method logic to the database query, that would improve performance even more. #butprobablyarefactoringforanotherday